### PR TITLE
refine asyncio call_exception_handler context type

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
I don't think we're ready to use PEP 655 in typeshed yet, unfortunately. #6301.

**Comment by sobolevn** [comment]
@JelleZijlstra looks like we can use intermediate `total=False` inheritance hack for this case.

File: stdlib/asyncio/base_events.pyi [hunk 1]
```diff
@@ -9,15 +9,26 @@ from asyncio.tasks import Task
...
 
 if sys.version_info >= (3, 7):
     from contextvars import Context
 
 _T = TypeVar("_T")
 _ProtocolT = TypeVar("_ProtocolT", bound=BaseProtocol)
-_Context = dict[str, Any]
+
+class _Context(TypedDict):
+    message: str
```
Comment (JelleZijlstra) [inline stdlib/asyncio/base_events.pyi:22]:
```suggestion
class _BaseContext(TypedDict):
    message: str
class _Context(_BaseContext, total=False):
```

File: stdlib/asyncio/base_events.pyi [hunk 2]
```diff
@@ -9,15 +9,26 @@ from asyncio.tasks import Task
...
+    exception: NotRequired[BaseException]
+    future: NotRequired[Future[Any]]
+    task: NotRequired[Task[Any]]
+    handle: NotRequired[Handle]
+    protocol: NotRequired[BaseProtocol]
+    transport: NotRequired[BaseTransport]
+    socket: NotRequired[socket]
+    asyncgen: NotRequired[AsyncGenerator[Any, Any]]
```
Comment (JelleZijlstra) [inline stdlib/asyncio/base_events.pyi:30]:
```suggestion
    exception: BaseException
    future: Future[Any]
    task: Task[Any]
    handle: Handle
    protocol: BaseProtocol
    transport: BaseTransport
    socket: socket
    asyncgen: AsyncGenerator[Any, Any]
```

File: stdlib/asyncio/events.pyi [hunk 3]
```diff
@@ -4,7 +4,7 @@ from _typeshed import FileDescriptorLike, Self
 from abc import ABCMeta, abstractmethod
 from socket import AddressFamily, SocketKind, _Address, _RetAddress, socket
 from typing import IO, Any, Awaitable, Callable, Coroutine, Generator, Sequence, TypeVar, Union, overload
-from typing_extensions import Literal
+from typing_extensions import Literal, NotRequired, TypedDict
 
 from .base_events import Server
```
Comment (JelleZijlstra) [inline stdlib/asyncio/events.pyi:9]:
```suggestion
from .base_events import Server, _Context
```

File: stdlib/asyncio/events.pyi [hunk 4]
```diff
@@ -18,7 +18,18 @@ if sys.version_info >= (3, 7):
...
+    message: str
+    exception: NotRequired[BaseException]
+    future: NotRequired[Future[Any]]
+    task: NotRequired[Task[Any]]
+    handle: NotRequired[Handle]
+    protocol: NotRequired[BaseProtocol]
+    transport: NotRequired[BaseTransport]
+    socket: NotRequired[socket]
+    asyncgen: NotRequired[AsyncGenerator[Any, Any]]
+
```
Comment (JelleZijlstra) [inline stdlib/asyncio/events.pyi:32]:
```suggestion
```

File: stdlib/asyncio/events.pyi [hunk 5]
```diff
@@ -4,7 +4,7 @@ from _typeshed import FileDescriptorLike, Self
 from abc import ABCMeta, abstractmethod
 from socket import AddressFamily, SocketKind, _Address, _RetAddress, socket
 from typing import IO, Any, Awaitable, Callable, Coroutine, Generator, Sequence, TypeVar, Union, overload
-from typing_extensions import Literal
+from typing_extensions import Literal, NotRequired, TypedDict
```
Comment (JelleZijlstra) [inline stdlib/asyncio/events.pyi:7]:
```suggestion
from typing_extensions import Literal, TypedDict
```

File: stdlib/asyncio/base_events.pyi [hunk 6]
```diff
@@ -9,15 +9,26 @@ from asyncio.tasks import Task
 from asyncio.transports import BaseTransport, ReadTransport, SubprocessTransport, WriteTransport
 from collections.abc import Iterable
 from socket import AddressFamily, SocketKind, _Address, _RetAddress, socket
-from typing import IO, Any, Awaitable, Callable, Coroutine, Generator, Sequence, TypeVar, Union, overload
-from typing_extensions import Literal
+from typing import IO, Any, AsyncGenerator, Awaitable, Callable, Coroutine, Generator, Sequence, TypeVar, Union, overload
+from typing_extensions import Literal, NotRequired, TypedDict
```
Comment (JelleZijlstra) [inline stdlib/asyncio/base_events.pyi:13]:
```suggestion
from typing_extensions import Literal, TypedDict
```

**Review by JelleZijlstra — COMMENTED** [review]
Let's see how much I can do in review suggestions

File: stdlib/asyncio/events.pyi [hunk 7]
```diff
@@ -4,9 +4,9 @@ from _typeshed import FileDescriptorLike, Self
 from abc import ABCMeta, abstractmethod
 from socket import AddressFamily, SocketKind, _Address, _RetAddress, socket
 from typing import IO, Any, Awaitable, Callable, Coroutine, Generator, Sequence, TypeVar, Union, overload
-from typing_extensions import Literal
+from typing_extensions import Literal, TypedDict
```
Comment (JelleZijlstra) [inline stdlib/asyncio/events.pyi:7]:
```suggestion
from typing_extensions import Literal
```

**Comment by JelleZijlstra** [comment]
Oops, I made pytype unhappy:
```
stdlib/asyncio/__init__.pyi (3.9): ParseError: 'total' allowed as classdef kwarg only for TypedDict subclasses
```
But I `_Context` is a TypedDict subclass. I filed google/pytype#1122.

File: stdlib/asyncio/base_events.pyi [hunk 8]
```diff
@@ -9,15 +9,28 @@ from asyncio.tasks import Task
...
     from contextvars import Context
 
 _T = TypeVar("_T")
 _ProtocolT = TypeVar("_ProtocolT", bound=BaseProtocol)
-_Context = dict[str, Any]
+
+class _BaseContext(TypedDict):
+    message: str
+
+class _Context(_BaseContext, total=False):
```
Comment (srittau) [inline stdlib/asyncio/base_events.pyi:24]: Looking forward to PEP 655.

`default_exception_handler()` also adds a `handle_trackback` field to the context:

https://github.com/python/cpython/blob/0a222db2bca63070f429c0e613707da1bdfaf0e0/Lib/asyncio/base_events.py#L1732-L1733

File: stdlib/asyncio/base_events.pyi [hunk 9]
```diff
@@ -9,15 +9,28 @@ from asyncio.tasks import Task
...
 
 if sys.version_info >= (3, 7):
     from contextvars import Context
 
 _T = TypeVar("_T")
 _ProtocolT = TypeVar("_ProtocolT", bound=BaseProtocol)
-_Context = dict[str, Any]
+
+class _BaseContext(TypedDict):
+    message: str
```
Comment (srittau) [inline stdlib/asyncio/base_events.pyi:22]: Is this really required? At least `default_exception_handler()` uses `.get()` to access it:

https://github.com/python/cpython/blob/0a222db2bca63070f429c0e613707da1bdfaf0e0/Lib/asyncio/base_events.py#L1719

File: stdlib/asyncio/base_events.pyi (hunk 8 above)
Comment (srittau) [reply stdlib/asyncio/base_events.pyi:24]: And `Handle._run()` adds `source_traceback` under some conditions:

https://github.com/python/cpython/blob/0a222db2bca63070f429c0e613707da1bdfaf0e0/Lib/asyncio/events.py#L92-L93

**Review by srittau — COMMENTED** [review]
+1 for for TypedDicts. A few things I noticed below.

**Review by hauntsaninja — COMMENTED** [review]
This is quite disruptive and I don't think we should merge. Because this type doesn't exist at runtime, I think it means in practice that any time you're not passing a dict literal, you'll get a type error, with no good way to refer to the type.

**Comment by AlexWaygood** [comment]
This no longer breaks pytype, so I've removed the "deferred" label. We now just need to decide whether it's a good idea or not.

File: stdlib/asyncio/events.pyi [hunk 10]
```diff
@@ -76,10 +76,8 @@ else:
 
 _T = TypeVar("_T")
 _ProtocolT = TypeVar("_ProtocolT", bound=BaseProtocol)
-_Context: TypeAlias = dict[str, Any]
 _ExceptionHandler: TypeAlias = Callable[[AbstractEventLoop, _Context], Any]
 _ProtocolFactory: TypeAlias = Callable[[], BaseProtocol]
```
Comment (AlexWaygood) [inline stdlib/asyncio/events.pyi:80]:
```suggestion
_ProtocolFactory: TypeAlias = Callable[[], BaseProtocol]
_SSLContext: TypeAlias = bool | None | ssl.SSLContext
```

Comment (AlexWaygood) [inline stdlib/asyncio/events.pyi:80]:
```suggestion
_ProtocolFactory: TypeAlias = Callable[[], BaseProtocol]
_SSLContext: TypeAlias = bool | None | ssl.SSLContext
```

Comment (AlexWaygood) [inline stdlib/asyncio/events.pyi:80]:
```suggestion
_ProtocolFactory: TypeAlias = Callable[[], BaseProtocol]

_SSLContext: TypeAlias = bool | None | ssl.SSLContext
```

**Comment by AlexWaygood** [comment]
I'm closing this PR: as @hauntsaninja says, this would be quite disruptive, and it would be complicated trying to access this `TypedDict` at runtime. I don't think we have enough consensus to make such a disruptive change; @srittau's review from February also has not been addressed.

The `asyncio` maintainers don't seem averse to typing, however: [some of the recent additions have been quite heavily typed](https://github.com/python/cpython/commit/f537b2a4fb86445ee3bd6ca7f10bc9d3a9f37da5). Perhaps a better way forward would be to file an issue over at CPython asking for this `TypedDict` to be added to the runtime. Then users would have an easy way of accessing the `TypedDict` at runtime, and so the concerns about disruption wouldn't be nearly so great.

**Pull request closed without merging**